system/security/plugins/jwtSecurity/jwtEndpoint.cpp

```diff
@@ -0,0 +1,192 @@
+/*##############################################################################
+
+    HPCC SYSTEMS software Copyright (C) 2020 HPCC Systems®.
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+############################################################################## */
+
+#include <curl/curl.h>
+#include <iostream>
+#include <string.h>
+
+#include "jlog.hpp"
+
+#include "nlohmann/json.hpp"
+
+#include "jwtEndpoint.hpp"
+
+struct MemoryStruct
+{
+    char*   memory;
+    size_t  size;
+    size_t  capacity;
+};
+
+void InitMemoryStructCapacity(MemoryStruct& mem, size_t initCapacity)
+{
+    mem.memory = static_cast<char*>(malloc(initCapacity));
+    mem.size = 0;
+    mem.capacity = initCapacity;
+}
+
+void FreeMemoryStruct(MemoryStruct& mem)
+{
+    if (mem.memory)
+    {
+        free(mem.memory);
+        mem.memory = nullptr;
+    }
+    mem.size = 0;
+    mem.capacity = 0;
+}
+
+size_t CaptureIncomingCURLReply(void* contents, size_t size, size_t nmemb, void* userp)
+{
+    size_t          incomingDataSize = size * nmemb;
+    MemoryStruct*   mem = static_cast<struct MemoryStruct*>(userp);
+    size_t          MAX_BUFFER_SIZE = 4194304; // 2^22
+
+    if ((mem->size + incomingDataSize) < MAX_BUFFER_SIZE)
+    {
+        if ((mem->size + incomingDataSize) > mem->capacity)
+        {
+            size_t  newCapacity = mem->capacity * 2;
+
+            // Keep doubling capacity until it is greater than what we need
+            while ((mem->size + incomingDataSize) > newCapacity)
+                newCapacity *= 2;
+
+            mem->memory = static_cast<char*>(realloc(mem->memory, newCapacity));
+            mem->capacity = newCapacity;
+        }
+
+        memcpy(&(mem->memory[mem->size]), contents, incomingDataSize);
+        mem->size += incomingDataSize;
+    }
+    else
+    {
+        // Signals an error to libcurl
+        incomingDataSize = 0;
+
+        PROGLOG("CJwtSecurityManager::CaptureIncomingCURLReply() exceeded buffer size %ld", MAX_BUFFER_SIZE);
+    }
+
+    return incomingDataSize;
+}
+
+std::string _tokenFromEndpoint(const std::string& jwtEndPoint, bool allowSelfSignedCert, const std::string& credentialsStr)
```

Comment:
Some inconsitent naming schemes here.  Lower case for extern?  Underscore for shared internal?  And uppercase for some other internals?
Should the internal functions be static?

Reply:
More copying and pasting from another project.  Thanks for catching these.  I've made the internal functions static as well, but I'm not sure if that makes much of a difference here.

Reply:
Thanks.  Different compilers have different default export semantics and trying to get in the habit of using static to avoid name collisions.